src/cmd/moby/run_qemu.go

```diff
@@ -53,78 +75,185 @@ func runQemu(args []string) {
 		log.Warnf("Both -iso and -uefi have been used")
 	}
 
-	// Before building qemu arguments, check qemu is in the $PATH
-	qemuBinPath := "qemu-system-" + *qemuArch
+	config := QemuConfig{
+		Prefix:   prefix,
+		ISO:      *qemuIso,
+		UEFI:     *qemuUEFI,
+		Kernel:   *qemuKernel,
+		GUI:      *qemuGUI,
+		DiskPath: *qemuDiskPath,
+		DiskSize: *qemuDiskSize,
+		FWPath:   *qemuFWPath,
+		Arch:     *qemuArch,
+		CPUs:     *qemuCPUs,
+		Memory:   *qemuMem,
+	}
+
+	config, qemuArgs := buildQemuCmdline(config)
+
+	var err error
+	if config.Containerized {
+		err = runQemuContainer(config, qemuArgs)
+	} else {
+		err = runQemuLocal(config, qemuArgs)
+	}
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+}
+
+func runQemuLocal(config QemuConfig, args []string) error {
+	if config.DiskPath != "" {
+		// If disk doesn't exist then create one
+		if _, err := os.Stat(config.DiskPath); os.IsNotExist(err) {
+			log.Infof("Creating new qemu disk [%s]", config.DiskPath)
+			qemuImgCmd := exec.Command(config.QemuImgPath, "create", "-f", "qcow2", config.DiskPath, config.DiskSize)
+			log.Debugf("%v\n", qemuImgCmd.Args)
+			if err := qemuImgCmd.Run(); err != nil {
+				return fmt.Errorf("Error creating disk [%s]:  %s", config.DiskPath, err.Error())
+			}
+		} else {
+			log.Infof("Using existing disk [%s]", config.DiskPath)
+		}
+	}
+
+	// Check for OVMF firmware before running
+	if config.UEFI {
+		_, err := os.Stat(config.FWPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return fmt.Errorf("File [%s] does not exist, please ensure OVMF is installed", config.FWPath)
+			}
+			return err
+		}
+	}
+
+	qemuCmd := exec.Command(config.QemuBinPath, args...)
+	// If verbosity is enabled print out the full path/arguments
+	log.Debugf("%v\n", qemuCmd.Args)
+
+	// If we're not using a seperate window then link the execution to stdin/out
+	if config.GUI != true {
+		qemuCmd.Stdin = os.Stdin
+		qemuCmd.Stdout = os.Stdout
+		qemuCmd.Stderr = os.Stderr
+	}
+
+	return qemuCmd.Run()
+}
+
+func runQemuContainer(config QemuConfig, args []string) error {
+	wd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	dockerArgs := []string{"run", "-i", "--rm", "-v", fmt.Sprintf("%s:%s", wd, "/tmp"), "-w", "/tmp"}
+
+	if config.KVM {
+		dockerArgs = append(dockerArgs, "--device", "/dev/kvm")
+	}
+
+	dockerPath, err := exec.LookPath("docker")
+	if err != nil {
+		return fmt.Errorf("Unable to find docker in the $PATH")
+	}
+
+	if config.DiskPath != "" {
+		// If disk doesn't exist then create one
+		if _, err = os.Stat(config.DiskPath); os.IsNotExist(err) {
```

Comment:
Also: could we share these common blocks of code between `runQemuContainer` and `runQemuLocal`?